try-runtime docs (#1056)
substrate-developer-hub/substrate-developer-hub.github.io: docs/knowledgebase/advanced/try-runtime.md

---
title: try-runtime
---

> The `try-runtime` tool is built to query a snapshot of runtime storage, using an in-memory-externalities to store state. In this way, 

Reviewer: Is there an explainer on externalities and the interface between runtime <-> client? Would be great to link with `in-memory-externalities`

Author: @emostov could you point me to some example of `in-memory-externalities`? 🙏🏻

Reviewer: I think it actually just reffers to `TestExternalities` - in memory reffering to the fact that in just exists in a hash map and is not rewritten to disk. I guess we can link just to https://substrate.dev/rustdocs/latest/sp_state_machine/struct.TestExternalities.html (but not worth it if we already link imo bc just more maintenance)

it enables runtime engineers to write tests for a specified runtime state.

> Being part of [`sc-cli`][sc-cli-rustdocs], it is designed to be used as a command line interface to specify 
> what block to capture and which pallets to run tests on.

Reviewer: ```suggestion
Being part of [`sc-cli`][sc-cli-rustdocs], it is designed to be used as a command line interface to specify at which block to query state.
```
Perhaps beyond scope, but could also be worth adding that when we query at a bock height we get the state after the block is applied against it with the given runtime.

In its simplest form, `try-runtime` is a tool that enables:

1. Connecting to a remote node and calling into some runtime API.

> 2. Taking some or all of the data for that node at a specified block.

Reviewer: ```suggestion
2. Scraping the specified state from a node at a given block.
```

3. Writing tests for that data.

> > **Note:** This tool is designed to be used for testing _before_ going to production.

Reviewer: ```suggestion
> **Note:** This tool is designed to be used for testing against real chain state _before_ going to production.
```

## Motivation

> The initial motivation for `try-runtime` came from the need to extend the capabilities of [`TestExternalities`][testextern-rustdocs] and [`BasicExternalities`][basicextern-rustdocs] 
> in such a way that it would be possible to use externalities to populate its memory with real chain's data. Since externalities provide the capability of storing an in-memory
> hash-map, extending these capabilities is achieved by simply using the key-value pairs from a node's RPCs.

Reviewer: ```suggestion
The initial motivation for `try-runtime` came from the need to test runtime changes against state from a real chain. Prior [`TestExternalities`][testextern-rustdocs] and [`BasicExternalities`][basicextern-rustdocs] existed for writing unit and integrated tests with mock data, but lacked an avenue to test against a chain's actual state. `try-runtime` extends [`TestExternalities`][testextern-rustdocs] and [`BasicExternalities`][basicextern-rustdocs] by scraping state (which is stored with key value pairs) via a node's RPC endpoints and inserting them into `TestExternalites`.
```

## How it works

> The `try-runtime` tool has its own implementation of externalities called [`remote_externalities`][remoteextern-rustdocs], with its hash-map containing:

Reviewer: Not clear to me if its important that its a hash-map. I think if we want to go into details we can describe it more generically as a key value store where most data is scale encoded. We can link to both scale encoding and a section on keys.
Some of these are a little dated but they are good explainers when it comes to keys and storage:
- https://www.shawntabrizi.com/substrate/querying-substrate-storage-via-rpc/
- https://www.shawntabrizi.com/substrate/substrate-storage-deep-dive/
- https://www.shawntabrizi.com/substrate/transparent-keys-in-substrate/

- a **Key**: the `Hash` of `hash(pallet_name) + hash(storage)`; and
- a **Value**: some SCALE encoded `Vec<u8>` Rust value. 

The diagram below illustrates the way externalities sits outside a compiled runtime as a means to capture 
the storage of that runtime. 

Storage externalities            |  Testing with externalities
:-------------------------:|:-------------------------:

> ![image](./../assets/advanced/try-runtime-ext-1.png)  |  ![image](./../assets/advanced/try-runtime-ext-2.png)

Reviewer: How can I view these pictures?

With `remote_externalities`, developers can capture some chain state and run tests on it. Essentially, `RemoteExterrnalities` will populate a `TestExternalities` with a real chain's data. 

In order to query state, `try-runtime` makes use of Substrate's RPCs, namely [`StateApi`][stateapi-rustdocs]. In particular:

> - [`storage`][stateapi-storage-rustdocs]: A method which returns a storage entry at a specific block’s state.

Reviewer: ```suggestion
- [`storage`][stateapi-storage-rustdocs]: A method which returns a storage value under the given key.
```
maybe? not sure exactly which rpc this is

### CLI interface

To use `try-runtime` from the command line, run your node with the `--features=try-runtime` flag. 

// TODO: basic commands / common combinations

The possible sub-commands include:

- **`on-runtime-upgrade`**: Executes "tryRuntime_on_runtime_upgrade" against the given runtime state.
- **`offchain-worker`**: Executes "offchainWorkerApi_offchain_worker" against the given runtime state.
- **`execute-block`**: Executes "core_execute_block" using the given block and the runtime state of the parent block.

For example, running `try-runtime` with the "on-runtime-upgrade" subcommand on a chain
running locally:

```bash
cargo run -- --release --features=try-runtime try-runtime on-runtime-upgrade live ws://localhost:9944
```

#### Other scenarios
Using it to re-execute code from a `ElectionProviderMultiPhase` off-chain worker:

```bash
cargo run -- --release \
--features=try-runtime \
try-runtime \
--execution Wasm \
--wasm-execution Compiled \
offchain-worker \
--header-at 0x491d09f313c707b5096650d76600f063b09835fd820e2916d3f8b0f5b45bec30 \
live \
-b 0x491d09f313c707b5096650d76600f063b09835fd820e2916d3f8b0f5b45bec30 \
-m ElectionProviderMultiPhase
ws//$HOST:9944
```

> 

Reviewer: Maybe its too trivial for this, but reminder users that after each (sub)command (`try-runtime` , `offchain-worker`, `live` etc) they can pass the `--help` flag to see the options. ATM the full command is a bit convoluted because you pass different flags to different subcommands, so a little tip might be nice

---
title: try-runtime
---

> The `try-runtime` tool is built to query a snapshot of runtime storage, using the externalities of a Substrate runtime to store state. In this way, 

Reviewer 2: ```suggestion
The `try-runtime` tool is built to query a snapshot of runtime storage, using an in-memory-externalities to store state. In this way, 
```

it enables runtime engineers to write tests for a specified runtime state.
Being part of [`sc-cli`][sc-cli-rustdocs], it is designed to be used as a command line interface to specify 
what block to capture and which pallets to run tests on.

In its simplest form, `try-runtime` is a tool that enables:

1. Connecting to a remote node and calling into some runtime API.
2. Taking some or all of the data for that node at a specified block.
3. Writing tests for that data.

> **Note:** This tool is designed to be used for testing _before_ going to production.

## Motivation

The initial motivation for `try-runtime` came from the need to extend the capabilities of [`TestExternalities`][testextern-rustdocs] and [`BasicExternalities`][basicextern-rustdocs] 

> in such a way that it would be possible to use externalities to populate its memory with real node data. Since externalities provide the capability of storing an in-memory

Reviewer 2: ```suggestion
in such a way that it would be possible to use externalities to populate its memory with real chain's data. Since externalities provide the capability of storing an in-memory
```

> hash-map, extending these capabilities is achieved by simply using the key-value pairs from a node's RPCs.

Reviewer 2: TODO: maybe link to which RPCs.

Reviewer: RPCs are getStorage, getKeysPaged, 

and we can also link some primers on how our keys work

## How it works
The `try-runtime` tool has its own implementation of externalities called [`remote_externalities`][remoteextern-rustdocs], with its hash-map containing:

- a **Key**: the `Hash` of `hash(pallet_name) + hash(storage)`; and

> - a **Value**: some SCALE encoded `Vec<u8>` Rust value. 

Reviewer 2: Some isseus here: 

`RemoteExternalities` is just a builder wrapper around `TestExternalities`.

Keys are not ALWAYS what you said, they sometimes are. How keys are generated is its own topic and you should link to it if it has a page.

The diagram below illustrates the way externalities sits outside a compiled runtime as a means to capture 
the storage of that runtime. 

Storage externalities            |  Testing with externalities
:-------------------------:|:-------------------------:
![image](./../assets/advanced/try-runtime-ext-1.png)  |  ![image](./../assets/advanced/try-runtime-ext-2.png)

> With `remote_externalities`, developers can capture some chain state and run tests on it. By contrast, `TestExternalities` only allows developers to 
> debug old state and `BasicExternalities` is designed only to be used in genesis. 

Reviewer 2: ```suggestion
With `remote_externalities`, developers can capture some chain state and run tests on it. Essentially, `RemoteExterrnalities` will populate a `TestExternalities` with a real chain's data. 
```

> In order to query state, `try-runtime` makes use of Substrate Core's [`StateApi`][stateapi-rustdocs]. In particular:

Reviewer 2: ```suggestion
In order to query state, `try-runtime` makes use of Substrate's RPCs, namely [`StateApi`][stateapi-rustdocs]. In particular:
```

- [`storage`][stateapi-storage-rustdocs]: A method which returns a storage entry at a specific block’s state.
- [`storage_key_paged`][stateapi-storage-keys-paged-rustdocs]: A method which returns the keys with prefix with pagination support.

// TODO: Add about frame::Executor

## Usage

The most common use case for `try-runtime` is with storage migrations and runtime upgrades.

### Calling into hooks from `OnRuntimeUpgrade`

> There's two ways of calling into `OnRuntimeUpgrade` hooks:

Reviewer 2: ```suggestion
There's two ways of defining an runtime upgrade hook in the runtime:
```

- **From inside a runtime**. For example:
    ```rust
    struct Custom;
    impl OnRuntimeUpgrade for Custom {
        fn on_runtime_upgrade() -> Weight { }
    }
    ```

- **From a pallet**. For example:
    ```rust
    #[pallet::hooks]
    fn on_runtime_upgrade() -> Weight { }
    ```

> To call into these hooks with `try-runtime`, a `feature` tag is used. This requires specifying a `features` 

Reviewer 2: ```suggestion
To call into these hooks with `try-runtime`, a cargo `feature` is used. This requires specifying a `features` 
```

dependency in your projects' relevant `Cargo.toml` files. 

> Refer to this how-to guide on integrating `try-runtime` to your project. (TODO)

> With this feature tag, `try-runtime` can call into `OnRuntimeUpgrade` hooks using the 

Reviewer 2: Please don't use the word tag. It is called a "cargo feature flag".

`#[cfg(feature = "try-runtime")]` macro. This macro must be 
declared before each function being tested in order for `try-runtime` to capture the relevant state 

> that its user specified. For example, in a storage migration, `try-runtime` can allow developers to 

Reviewer 2: This section is a bit superfolous. I am not sure what 

> to capture the relevant state that its user specified

 means. 

I think what you are trying to say is: 

- there are two ways to implement on-runtime upgrade. These are methods are `OnRuntimeUpgrade` trait.
- if the feature is enabled, you can implement two additional methods as well.
- if you do so, these methods are called when you use `try-runtime on-runtime-upgrade`.

Using the [`frame_executive::Executive`][executive-rustdocs] struct, these helper functions in action would look like:

```rust
pub struct CheckTotalIssuance;
impl OnRuntimeUpgrade for CheckTotalIssuance {
    #[cfg(feature = "try-runtime")]
    fn post_upgrade() {
        // iterate over all accounts, sum their balance and ensure that sum is correct.
    }
}

pub struct EnsureAccountsWontDie;
impl OnRuntimeUpgrade for EnsureAccountsWontDie {
    #[cfg(feature = "try-runtime")]
    fn pre_upgrade() {
        let account_count = frame_system::Accounts::<Runtime>::iter().count();
        Self::set_temp_storage(account_count, "account_count");
    }

    #[cfg(feature = "try-runtime")]
    fn post_upgrade() {
        // ensure that this migration doesn't kill any account.
        let post_migration = frame_system::Accounts::<Runtime>::iter().count();
        let pre_migration = Self::get_temp_storage::<u32>("account_count");
        ensure!(post_migration == pre_migration, "error ...");
    }
}

pub type CheckerMigrations = (EnsureAccountsWontDie, CheckTotalIssuance);
pub type Executive = Executive<_, _, _, _, (CheckerMigrations)>;
```

### CLI interface

> To use `try-runtime` from the command line, run your node with the `--features=try-runtime` flag. 

Reviewer 3: It might be worth mentioning that `cargo run` is used here or specifying that the `try-runtime` subcommand is used when invoking the binary directly.

Author: > > Thanks for putting this together.
> > I'm interested in using `try-runtime` to ensure that our upcoming migrations go as expected. While there is a lot that can be done with the additional `pre_upgrade()` and `post_upgrade()` hooks, I would like to use `try-runtime` to begin > > Is this possible with `try-runtime`? If so, I'd love to see it mentioned in documentation.
> 
> Ah, it was this that I was thinking of: https://github.com/maxsam4/fork-off-substrate
> 
> I'm yet to play with it, but it looks incredibly useful. It might be worth mentioning in `try-runtime`'s documentation.

Hi Stephen - thanks for your contribution and comments here, I'll make an attempt to make this a little clearer in the article here. The answer is that that `try-runtime` seems like a useful tool in the scenario you're describing. Unlike 'fork-off-substrate', you can write unit tests in the runtime of your chain or specific pallets in `on_runtime_upgrade`.

- [`storage_key_paged`][stateapi-storage-keys-paged-rustdocs]: A method which returns the keys with prefix with pagination support.

## Usage

The most common use case for `try-runtime` is with storage migrations and runtime upgrades. 

> **Tip:** Combine `try-runtime` with `fork-off-substrate` to test your chain before production. Use `try-runtime` to test your 
> chain's migration and its pre and post states. Then, use [`fork-off-substrate`][fork-off-gh] if you want to check that block 
> production continues fine after the migration, and do some other arbitrary testing. 

### Calling into hooks from `OnRuntimeUpgrade`
By default, there are two ways of defining a runtime upgrade in the runtime. The `OnRuntimeUpgrade` trait provides the [different methods][onruntimeupgrade-method-rustdocs] to achieve this.

- **From inside a runtime**. For example:
    ```rust
    struct Custom;
    impl OnRuntimeUpgrade for Custom {
        fn on_runtime_upgrade() -> Weight { }
    }
    ```

- **From inside a pallet**. For example:
    ```rust
    #[pallet::hooks]
    fn on_runtime_upgrade() -> Weight { }
    ```

> To call into these hooks with `try-runtime`, a cargo `feature` is used. This requires specifying a `features` 

Reviewer 2: This is false. To call into these hooks, you don't need a feature. To call into the additional hooks (e.g. `pre_upgrade`) you need the feature.

> To call into these hooks with `try-runtime`, a cargo `feature` is used. This requires specifying a `features` 
> dependency in your projects' relevant `Cargo.toml` files. With this cargo feature flag enabled, `try-runtime` can call into `OnRuntimeUpgrade` hooks using the 
> `#[cfg(feature = "try-runtime")]` macro which gives it access to the `OnRuntimeUpgrade` methods.
> 
> > Refer to this how-to guide on integrating `try-runtime` to your project. _(TODO)_
> 
> For example, in a storage migration, `try-runtime` can allow developers to 
> examine state both before and after a runtime upgrade:
> 
> ```rust
>     #[cfg(feature = "try-runtime")]
>     fn pre_upgrade() -> Result<(), &'static str> { Ok(()) }
> 
>     #[cfg(feature = "try-runtime")]
>     fn post_upgrade() -> Result<(), &'static str> { Ok(()) }
> ```
> 
> > See how the Executive pallet implements the `try_runtime_upgrade` method [here][executive-example-frame] to 
> > aggregate all `OnRuntimeUpgrade` calls from a runtime and return a `Weight`, which could be used for testing.  

Reviewer 2: ```suggestion

These hooks will specify _what should happen upon a runtime upgrade_. For testing purposes, we prefer having hooks that allow us to inspect the state _before_ and _after_ a runtime upgrade as well. 

These hooks are not available by default, and are only available under a specific feature flag, named `try-runtime`. 

> Refer to this how-to guide on integrating `try-runtime` to your project. _(TODO)_

The new hooks are as follows:

```rust
    #[cfg(feature = "try-runtime")]
    fn pre_upgrade() -> Result<(), &'static str> { Ok(()) }

    #[cfg(feature = "try-runtime")]
    fn post_upgrade() -> Result<(), &'static str> { Ok(()) }
```

> See how the Executive pallet implements the `try_runtime_upgrade` method [here][executive-example-frame] to 
> aggregate all `OnRuntimeUpgrade` calls from a runtime and return a `Weight`, which could be used for testing.  
```

- [`storage`][stateapi-storage-rustdocs]: A method which returns a storage value under the given key.
- [`storage_key_paged`][stateapi-storage-keys-paged-rustdocs]: A method which returns the keys with prefix with pagination support.

## Usage

The most common use case for `try-runtime` is with storage migrations and runtime upgrades. 

> **Tip:** Combine `try-runtime` with `fork-off-substrate` to test your chain before production. Use `try-runtime` to test your 
> chain's migration and its pre and post states. Then, use [`fork-off-substrate`][fork-off-gh] if you want to check that block 

> > production continues fine after the migration, and do some other arbitrary testing. 

Reviewer 3: :+1: thanks!
